lib/include/openamp/rpmsg_rpc_client_server.h

```diff
@@ -0,0 +1,198 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef RPMSG_RPC_CLIENT_SERVER_H
+#define RPMSG_RPC_CLIENT_SERVER_H
+
+#include <metal/mutex.h>
+#include <openamp/open_amp.h>
+#include <stdint.h>
+
+#if defined __cplusplus
+extern "C" {
+#endif
+
+/* RPMSG_BUFFER_SIZE = 512
+ * sizeof(struct rpmsg_hdr) = 16
+ * RPMSG_BUFFER_SIZE - sizeof(struct rpmsg_hdr) - 1 = 495
+ * Aligning to 64 bits -> 488UL
+ */
+#define MAX_BUF_LEN	488UL
+#define MAX_FUNC_ID_LEN sizeof(unsigned long int)
+#define TABLE_SIZE	10
+struct rpmsg_rpc_data;
+
+typedef int (*rpmsg_rpc_poll)(void *arg);
+typedef void (*rpmsg_rpc_shutdown_cb)(struct rpmsg_rpc_data *rpc);
+typedef void (*app_cb)(struct rpmsg_rpc_data *rpc, void *data, size_t len);
+
+struct rpmsg_rpc_syscall_cb {
+	int (*cb_function)(void *data, struct rpmsg_endpoint *ept);
+};
+
+/**
+ * struct rpmsg_rpc_services - table for services
+ *
+ * @id: service id
+ * @respcb: id callback
+ *
+ */
+struct rpmsg_rpc_services {
+	uint32_t id;
+	struct rpmsg_rpc_syscall_cb respcb;
+};
+
+/**
+ * struct rpmsg_rpc_data - remote procedure call for global data
+ *
+ * RPMsg RPC will send request to remote and
+ * wait for callback.
+ *
+ * @ept: rpmsg_endpoint structure
+ * @nacked: flag
+ * @reqbuf: pointer to request structure
+ * @respbuf: pointer to response structure
+ * @respbuf_len: size of response buffer
+ * @cb: app callback function
+ * @poll: poll function
+ * @poll_arg: pointer to poll function argument
+ * @shutdown_cb: shutdown callback function
+ * @lock: mutex element
+ * @buflock: metal spinlock
+ *
+ */
+struct rpmsg_rpc_data {
+	struct rpmsg_endpoint ept;
+	atomic_int nacked;
+	void *reqbuf;
+	void *respbuf;
+	size_t respbuf_len;
+	app_cb cb;
+	rpmsg_rpc_poll poll;
+	void *poll_arg;
+	rpmsg_rpc_shutdown_cb shutdown_cb;
+	metal_mutex_t lock;
+	struct metal_spinlock buflock;
+};
+
+/**
+ * rpmsg_rpc_client_release - release RPMsg remote procedure call
+ *
+ * This function is to release remoteproc procedure call
+ * global data.
+ *
+ * @rpc: pointer to the global remote procedure call data
+ *
+ */
+void rpmsg_rpc_client_release(struct rpmsg_rpc_data *rpc);
+
+/**
+ * rpmsg_rpc_client_init - initialize RPMsg remote procedure call
+ *
+ * This function is to initialize the remote procedure call
+ * global data. RPMsg RPC will send request to remote and
+ * wait for callback.
+ *
+ * @rpc: pointer to the global remote procedure call data
+ * @rdev: pointer to the rpmsg device
+ * @ept_name: name of the endpoint used by RPC
+ * @ept_addr: address of the endpoint used by RPC
+ * @ept_raddr: remote address of the endpoint used by RPC
+ * @poll_arg: pointer to poll function argument
+ * @poll: poll function
```

Comment:
poll field seems to me uselless, can be full handled in application part. can be also removed from rpmsg_rpc_data struct